examples/nivc_dynamic_phi.rs

```diff
@@ -0,0 +1,360 @@
+use std::marker::PhantomData;
+
+use bellpepper::gadgets::{blake2s::blake2s, multipack::pack_bits, sha256::sha256};
+use bellpepper_core::{boolean::Boolean, num::AllocatedNum, ConstraintSystem, SynthesisError};
+use blake2s_simd::Params;
+use ff::{Field, PrimeField, PrimeFieldBits};
+use rand::Rng;
+use rand_core::OsRng;
+use sha2::{Digest, Sha256};
+
+use nova_snark::{
+  parameters::PublicParams,
+  supernova::{NonUniformCircuit, RecursiveSNARK},
+  traits::{
+    circuit_supernova::{StepCircuit, TrivialSecondaryCircuit},
+    Group,
+  },
+};
+
+const NUM_STEPS: usize = 10;
+const NUM_BYTES: usize = 10;
+
+#[derive(Clone, Debug)]
+struct SHACircuit<F: PrimeField> {
+  next_pc: bool,
+  _p: PhantomData<F>,
+}
+
+impl<F: PrimeField> SHACircuit<F> {
+  fn new(next_pc: bool) -> Self {
+    Self {
+      next_pc,
+      _p: PhantomData,
+    }
+  }
+}
+
+impl<F: PrimeField + PrimeFieldBits> StepCircuit<F> for SHACircuit<F> {
+  fn arity(&self) -> usize {
+    1
+  }
+
+  fn synthesize<CS: ConstraintSystem<F>>(
+    &self,
+    cs: &mut CS,
+    _pc: Option<&AllocatedNum<F>>,
+    z: &[AllocatedNum<F>],
+  ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+    let preimage = &z[0];
+
+    let preimage_bits = preimage.to_bits_le_strict(cs.namespace(|| "sha_preimage_bits"))?;
+
+    let (preimage_bits, _) = preimage_bits.split_at(8 * NUM_BYTES);
+
+    let preimage_bits = preimage_bits
+      .chunks(8)
+      .flat_map(|chunk| {
+        let mut chunk = chunk.to_vec();
+        chunk.reverse();
+        chunk
+      })
+      .collect::<Vec<_>>();
+
+    let digest_bits = sha256(cs.namespace(|| "sha256"), preimage_bits.as_slice())?;
+
+    let (digest_bits, _) = digest_bits.split_at(8 * NUM_BYTES);
+
+    let next_pc_bit = &digest_bits[7];
+
+    let digest_bits = digest_bits
+      .chunks(8)
+      .flat_map(|chunk| {
+        let mut chunk = chunk.to_vec();
+        chunk.reverse();
+        chunk
+      })
+      .collect::<Vec<Boolean>>();
+
+    let digest = pack_bits(cs.namespace(|| "digest_from_bits"), digest_bits.as_slice())?;
+
+    let new_pc = AllocatedNum::alloc(cs.namespace(|| "next_pc"), || {
+      if self.next_pc {
+        Ok(F::ONE)
+      } else {
+        Ok(F::ZERO)
+      }
+    })?;
+
+    cs.enforce(
+      || "enforce new_pc is first bit",
+      |lc| lc + CS::one(),
+      |lc| lc + new_pc.get_variable(),
+      |_| next_pc_bit.lc(CS::one(), F::ONE),
+    );
+
+    Ok((Some(new_pc), vec![digest]))
+  }
+}
+
+#[derive(Clone, Debug)]
+struct BlakeCircuit<F: PrimeField> {
+  next_pc: bool,
+  _p: PhantomData<F>,
+}
+
+impl<F: PrimeField> BlakeCircuit<F> {
+  fn new(next_pc: bool) -> Self {
+    Self {
+      next_pc,
```

Comment:
Same comments/questions about the need for this hint as in the sha circuit above.